dtm0/service.c

```diff
@@ -453,223 +484,602 @@ M0_INTERNAL bool m0_dtm0_in_ut(void)
+	m0_long_lock_fini(&proc->dop_llock);
+}
 
-	M0_PRE(reqh_service != NULL);
-	service = container_of(reqh_service, struct m0_dtm0_service,
-			       dos_generic);
+static void dtm0_service_conns_term(struct m0_dtm0_service *service)
+{
+	struct dtm0_process *process;
+	int                  rc;
 
 	M0_ENTRY("dtms=%p", service);
 
 	while ((process = dopr_tlist_pop(&service->dos_processes)) != NULL) {
 		rc = dtm0_process_disconnect(process);
-		M0_ASSERT_INFO(rc == 0, "TODO: Disconnect failures"
-			       " are not handled yet.");
-		dtm0_process__ha_state_unsubscribe(process);
-		dopr_tlink_fini(process);
+		M0_ASSERT_INFO(rc == 0, "Failed to disconnect from %p?",
+			       process);
+		dtm0_process_fini(process);
 		m0_free(process);
 	}
 
 	M0_LEAVE();
 }
 
+static int find_or_add(struct m0_dtm0_service *dtms,
+		       const struct m0_fid    *tgt,
+		       struct dtm0_process   **out)
+{
+	struct dtm0_process *process;
+	int                  rc;
+
+	M0_ENTRY();
+	M0_PRE(m0_mutex_is_locked(&dtms->dos_generic.rs_mutex));
+
+	process = dtm0_service_process__lookup(&dtms->dos_generic, tgt);
+	if (process == NULL) {
+		rc = M0_ALLOC_PTR(process) == NULL ?
+			M0_ERR(-ENOMEM) :
+			dtm0_process_init(process, dtms, tgt);
+		if (rc != 0)
+			m0_free(process);
+	}
+
+	if (rc == 0) {
+		M0_POST(process != NULL);
+		*out = process;
+	}
+	return M0_RC(rc);
+}
+
+enum drlink_fom_state {
+	DRF_INIT = M0_FOM_PHASE_INIT,
+	DRF_DONE = M0_FOM_PHASE_FINISH,
+	DRF_LOCKING = M0_FOM_PHASE_NR,
+	DRF_CONNECTING,
+	DRF_SENDING,
+	DRF_WAITING_FOR_REPLY,
+	DRF_FAILED,
+	DRF_NR,
+};
+
+static struct m0_sm_state_descr drlink_fom_states[] = {
+	/* terminal states */
+	[DRF_INIT] = {
+		.sd_name      = "DRF_INIT",
+		.sd_allowed   = M0_BITS(DRF_LOCKING, DRF_FAILED),
+		.sd_flags     = M0_SDF_INITIAL,
+	},
+	[DRF_DONE] = {
+		.sd_name      = "DRF_DONE",
+		.sd_allowed   = 0,
+		.sd_flags     = M0_SDF_TERMINAL,
+	},
+
+	/* failure states */
+	[DRF_FAILED] = {
+		.sd_name      = "DRF_FAILED",
+		.sd_allowed   = M0_BITS(DRF_DONE),
+		.sd_flags     = M0_SDF_FAILURE,
+	},
+
+	/* intermediate states */
+#define _ST(name, allowed)            \
+	[name] = {                    \
+		.sd_name    = #name,  \
+		.sd_allowed = allowed \
+	}
+	_ST(DRF_LOCKING,           M0_BITS(DRF_CONNECTING,
+					   DRF_SENDING,
+					   DRF_FAILED)),
+	_ST(DRF_CONNECTING,        M0_BITS(DRF_SENDING,
+					   DRF_FAILED)),
+	_ST(DRF_SENDING,           M0_BITS(DRF_DONE,
+					   DRF_WAITING_FOR_REPLY,
+					   DRF_FAILED)),
+	_ST(DRF_WAITING_FOR_REPLY, M0_BITS(DRF_DONE,
+					   DRF_FAILED)),
+#undef _ST
+};
+
+const static struct m0_sm_conf drlink_fom_conf = {
+	.scf_name      = "m0_dtm0_drlink_fom",
+	.scf_nr_states = ARRAY_SIZE(drlink_fom_states),
+	.scf_state     = drlink_fom_states,
+};
+
+static struct drlink_fom *item2drlink_fom(struct m0_rpc_item *item)
+{
+	return m0_rpc_item_to_fop(item)->f_opaque;
+}
+
+static void dtm0_rlink_rpc_item_reply_cb(struct m0_rpc_item *item)
```

Comment:
it's easier to subscribe on item->ri_sm.chan, still let's postpone this change.

Reply:
Yes, let's do that later. co_op is the simplest solution at the moment.